crates/service/src/prelude/storage/mod.rs

```diff
@@ -0,0 +1,81 @@
+mod dense;
+mod sparse;
+
+pub use dense::DenseMmap;
+pub use sparse::SparseMmap;
+
+use crate::index::IndexOptions;
+use crate::prelude::*;
+use std::borrow::Cow;
+use std::path::Path;
+
+pub trait Ram {
+    type Element: Copy + bytemuck::Pod;
+
+    fn dims(&self) -> u16;
+    fn len(&self) -> u32;
+    fn content(&self, i: u32) -> &[Self::Element];
+    fn payload(&self, i: u32) -> Payload;
+}
+
+pub trait Storage {
+    type Element: Copy + bytemuck::Pod;
+    type Scalar: Copy;
+    type Vector: Vector<Element = Self::Element>;
```

Comment:
Move these assiocated types to trait `G`.

Reply:
completed in new commit